### ISSUE TITLE
Cannot create record set from EO query when exchangeDate is set 

### ISSUE THREAD

**Issue by grantfitzsimmons** [specify/specify7#2503]
```
(_mysql_exceptions.OperationalError) (1292, "Truncated incorrect datetime value: '10/17/2016'") [SQL: "INSERT INTO recordsetitem (`RecordId`, `RecordSetID`) SELECT DISTINCT exchangeout.`ExchangeOutID` AS `exchangeout_ExchangeOutID`, %s AS param_1 \nFROM exchangeout LEFT OUTER JOIN agent AS agent_1 ON agent_1.`AgentID` = exchangeout.`SentToOrganizationID` LEFT OUTER JOIN address AS address_1 ON agent_1.`AgentID` = address_1.`AgentID` LEFT OUTER JOIN shipment AS shipment_1 ON exchangeout.`ExchangeOutID` = shipment_1.`ExchangeOutID` \nWHERE exchangeout.`DivisionID` = %s AND (address_1.`Address` LIKE concat(concat('%%', %s), '%%')) AND exchangeout.`ExchangeDate` = %s"] [parameters: (9752, 2, 'beal', '10/17/2016')] (Background on this error at: http://sqlalche.me/e/e3q8)
```
![chrome_fRVZy6wsIb](https://user-images.githubusercontent.com/37256050/204057648-0c7ff791-ec5f-4635-b231-90cdf504fed6.gif)

This occurs when trying to create a record set from this exchange out query.

**If the `exchangeDate` is not set to a date (in this case, 10/17/2016), you can create one successfully!**

https://umich-testability.test.specifysystems.org/specify/query/117/

This error happens on `edge` and `testability`

**Comment by maxxxxxdlp** [comment]
I didn't look at the code, but it appears that the back-end is using the formatted date (using the formatter defined in remote prefs), where as it is supposed to use the mysql date format when communicating to mysql

**Comment by realVinayak** [comment]
@grantfitzsimmons If you remove the first field (description), and add it again (which shouldn't change the query or the results) and then run the Create Record Set then it works fine. @maxxxxxdlp  is right about the dates. The first object is the one that works (if you remove a field and add it back) and second one is the one that doesn't work. Notice the date format in startvalue field. 

**This works**
```
{
    "tablelist": "40",
    "stringid": "40.ExchangeOut.exchangeDate",
    "fieldname": "exchangeDate",
    "isrelfld": false,
    "sorttype": 0,
    "position": 7,
    "isdisplay": true,
    "operstart": 1,
    "startvalue": "2016-10-17",
    "isnot": false
}

```

**This doesn't work**
```
{
    "id": 1682,
    "allownulls": null,
    "alwaysfilter": false,
    "columnalias": "exchangeDate",
    "contexttableident": 40,
    "endvalue": "",
    "fieldname": "exchangeDate",
    "formatname": null,
    "isdisplay": true,
    "isnot": false,
    "isprompt": true,
    "isrelfld": false,
    "operend": null,
    "operstart": 1,
    "position": 8,
    "sorttype": 0,
    "startvalue": "10/17/2016",
    "stringid": "40.exchangeout.exchangeDate",
    "tablelist": "40",
    "timestampcreated": "2017-03-21T11:43:49",
    "timestampmodified": "2017-03-21T11:43:49",
    "version": 0,
    "createdbyagent": "/api/specify/agent/2/",
    "modifiedbyagent": null,
    "query": "/api/specify/spquery/117/",
    "mappings": "/api/specify/spexportschemaitemmapping/?queryfield=1682",
    "resource_uri": "/api/specify/spqueryfield/1682/"
}
```

**Comment by maxxxxxdlp** [comment]
Fixed in e0aa78dbe. Needs testing

**Comment by maxxxxxdlp** [comment]
@realVinayak While QA people do manual testing, please also add automated tests for SpQueryField

**Comment by maxxxxxdlp** [comment]
Relevant conversation about this between me and Jason:

TLDR: Sp6 saves dates in the format you configured for that database (in this case `10/17/2016`) - a really bad idea as admins could always reconfigure that forma. Sp7, like any sane software, stores date in a machine-standart format of `2016-10-17`.

Max
:globe_with_meridians:  [2:42 PM](https://specifydev.slack.com/archives/D040LK4L4QM/p1669668148811109)
Can you try creating a query with a date filter in sp6 and see that it saves the date in the same format as sp7?


Jason Melton
  [4:36 PM](https://specifydev.slack.com/archives/D040LK4L4QM/p1669674972737549)
Ok, so if you meant the date format, Specify 7 does respect the date filter--- as far as the value in the database is concerned.
In 7, the correct date format shows in the results of the query builder. However, in other places of the UI (Such as the partial date fields), Specify uses the MM/DD/YYYY format (this may just be browser/region dependent).
For instance, when a new record is saved (or old record is viewed), on the form it is displayed in the MM/DD/YYYY format, but it saves in the database (and prints out in the Query Builder) with the correct format.
I will get back to you briefly if you meant a date constraint within the query itself- I am waiting for docker to finish building.


Jason Melton
  [5:00 PM](https://specifydev.slack.com/archives/D040LK4L4QM/p1669676422168459)
Ok, constraints regarding the date of a query itself (i.e. Between) was getting some strange behavior. Most notably, I changed the date format in 6 to use DD.MM.YYYY  and setup a query to return results between 31.10.2022 and 01.11.2022 .
When ported to Specify 7 (date format MM/DD/YYYY), the same query was running between 10/01/2022  and 01/11/2022.
When hovering over one of the Partial Date UI components in 7, the tooltip is Required Format: DD.MM.YYYY.


Max
:globe_with_meridians:  [6:00 PM](https://specifydev.slack.com/archives/D040LK4L4QM/p1669680000192579)
I see!
What about the database? Can you take a look at the SpQuery and SpQueryField tables to see what value does sp6 and sp7 put inot the SpQueryField.startValue field when you save a query?


Jason Melton
  [9:21 AM](https://specifydev.slack.com/archives/D040LK4L4QM/p1669735315976819)
Here are the results (startvalue field) when I recreated and saved the same query in Sp 7:
Specify 6:  31.10.2022,01.11.2022
Specify 7:  2022-10-31,2022-11-01


Max
:globe_with_meridians:  [10:17 AM](https://specifydev.slack.com/archives/D040LK4L4QM/p1669738642121329)
Different formats - that's bad
[10:17](https://specifydev.slack.com/archives/D040LK4L4QM/p1669738661743919)
If you save a query in sp6 and open it in sp7, would the date display correctly in the UI? would the query run?
[10:17](https://specifydev.slack.com/archives/D040LK4L4QM/p1669738674537209)
Does it also work the other way? (create a query in sp7 and open it in sp6)


Jason Melton
  [10:51 AM](https://specifydev.slack.com/archives/D040LK4L4QM/p1669740663674099)
That was some of the testing I did with this earlier: when a query is created in sp6 and ran in sp7, the format is correct in the results, but every Partial Date UI component uses mm/dd/yyyy format. The user only ever sees their format in the QueryBuilder results.
Exporting the query from Sp7 to Sp6 (which has the aforementioned altered format) thankfully causes no fatal errors: the dates in the Query are formatted in mm/dd/yyyy , yet the Query still runs and returns the correct results (formatted in the right way)


Max
:globe_with_meridians:  [10:53 AM](https://specifydev.slack.com/archives/D040LK4L4QM/p1669740783301179)
Weird. So the formats are different, but it just works? Nice!
[10:53](https://specifydev.slack.com/archives/D040LK4L4QM/p1669740832448569)
The front-end query builder expects the date in one format, but has a fallback for other date formats - that's probably why it works in the UI. and since UI send the query to the back-end, back-end receives the query with a correct date format

**Comment by maxxxxxdlp** [comment]
@grantfitzsimmons is this a priority enough to include in 7.8?
Fixing shouldn't be difficult, but good amount of testing would be required.

**Comment by grantfitzsimmons** [comment]
Not a priority for 7.8. We can solve at a later date.

**Comment by grantfitzsimmons** [comment]
Looks like a good candidate for a 7.9.x point release
